*: add restriction to running DDL with internal executors

ZhouXing19 opened the pull request:
The current internal executor has its lifecycle, which makes it erroneous
when being used to execute DDL statements if under an outer txn. 

In this commit, we 
1. Migrated the existing DDLs with internal executor with not-nil txn to either 
`descs.CollectionFactory.TxnWithExecutor()` or `planner.WithInternalExecutor()`.
Only internal executors inited via these 2 interfaces are bounded with txn-related
metadata, and hence are allowed to run DDLs in a transactional manner.
2. Added a restriction for running DDLs with internal executor only if it's bound with
txn-related metadata.

fixes #87281

Release justification: bug fix for the internal executor
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86334)
<!-- Reviewable:end -->

ajwerner commented:
What should we do if the internal executor was constructed with a transaction and then a non-nil transaction is passed in?

ZhouXing19 commented:
> What should we do if the internal executor was constructed with a transaction and then a non-nil transaction is passed in?

In the old design (prior to https://github.com/cockroachdb/cockroach/pull/82477), we don't construct an internal executor with a txn. Each time a not-nil txn passed in with a statement, we construct a conn executor with that txn. The internal executor is not bound to any txn at all.

In the new interfaces, we actually didn't truly used the txn bounded to the internal executor (`ie.extraTxnState.txn`), except for `InternalExecutor.commitTxn()`, which is only used under the `collectionFactory`. So essentially, the txn that the conn executor is constructed with is still the one from the Ex functions.

Ideally, we should disallow passing a txn through the Ex functions, and always use `ie.extraTxnState.txn`. But the prerequisite is that all internal executor usages are using the new interface, and this will involve a big change. So we didn't include it in #82477.

ZhouXing19 commented:
I don't think the current simple `tree.CanModifySchema(parsed.AST) && txn != nil` check will work. I changed to check if the internal executor is bound to an explicit txn. But in this case, this check is only effective with IE inited via the new interfaces.

ajwerner commented:
Can you say more about why the check does not work?

ajwerner commented:
It feels like we need it to be the case that: 
1) If you use the new interface and the user passes a different transaction, that should definitely be an error. I don't know what we should do if the user passes nil, but my instinct is that for now that should also be an error
2) If you try to do a DDL and you didn't construct the transaction in the right way with a bound collection etc, that should be an error.

ZhouXing19 commented:
> Can you say more about why the check does not work?

Maybe I gave the conclusion too fast -- it was because I saw it broke a great number of tests in the CI. But they were broken because it couldn't run [`create-file-table`](https://github.com/cockroachdb/cockroach/blob/79ef82000093190c12991edda480866154536b85/pkg/cloud/userfile/filetable/file_table_read_writer.go#L828), and [clean up temporary type](https://github.com/cockroachdb/cockroach/blob/79ef82000093190c12991edda480866154536b85/pkg/sql/temporary_schema.go#L388). Maybe we should just pass nil txn for them? 

> It feels like we need it to be the case that ...

Yeah, it makes sense, will add them as well. (Though I have a stronger feeling that the new interface makes everything messier now 😥)

ZhouXing19 commented:
> If you try to do a DDL and you didn't construct the transaction in the right way with a bound collection etc, that should be an error.

If we want to apply this restriction, I think we still need to find out all existing DDL executions with not-nil txn, and either let them run with `collectionFactory.TxnWithExeutor()` or with nil txn. It turns out that except for those listed in #76764, there're [more of them](https://github.com/cockroachdb/cockroach/commit/b8caec51d2c0ef12c7b21bcbd5e3e4a9f3d0bf4b). I'm trying to change them to use nil txn for now and see what will happen ...

ZhouXing19 commented:
I've changed this PR to refactor the DDL use cases with `descs.TxnWithExecutor()`, which is broken and relies on https://github.com/cockroachdb/cockroach/pull/86427. This PR may also overlap with https://github.com/cockroachdb/cockroach/pull/86433, so I'll rebase from these 2 afterward.

ZhouXing19 committed:
*: infuse `CollectionFactory` for storage and backup

This is part of the migration of existing DDL statement with internal executor
to `descs.CollectionFactory.TxnWithExecutor()`. DDL statements should only be
run with an internal executor that's created via this function.

Release justification: Low risk, high benefit changes to existing functionality
Release note: none

ZhouXing19 committed:
filetable: use `CollectionFactory.TxnWithExecutor()` for DDL statements

fixes #76764

Release justification: Low risk, high benefit changes to existing functionality
Release note: none

ZhouXing19 committed:
sql: migreate authorization_test to `CollectionFactory.TxnWithExecutor()` for DDLs

This is part of the project to migrate existing DDL statements running with
an internal executor to `descs.CollectionFactory()`. DDLs are only allowed
to run with internal executor inited via this function.

Release justification: Low risk, high benefit changes to existing functionality
Release note: None

ZhouXing19 committed:
sql: migrate create_external_connection to `planner.WithInternalExecutor()` for DDLs

This commit is to migrate the existing DDLs to using
`planner.WithInternalExecutor()`. DDLs with internal executors are only allowed
if the latter is bounded with txn-realated metadata.

Release justification: Low risk, high benefit changes to existing functionality
Release note: none

ZhouXing19 committed:
backupccl: migrate job restoration to `CollectionFactory.TxnWithExecutor()` for DDLs

This commit is part of the project to migrate DDLs running with internal executor
with the correct interface. DDLs are only allowed to run with internal executor
that is bound to txn-related metadata.

Release justification: Low risk, high benefit changes to existing functionality
Release note: none

ZhouXing19 committed:
backupccl: refactor `cleanupTempSystemTables()`

We stripped `txn` from the parameter list in `cleanupTempSystemTables()`.
It was run with not-nil txn by mistake, which is a mis-usage of running
internal executor with DDLs.

Release justification: bug fix
Release note: none

ajwerner commented on pkg/sql/catalog/collectionfactory/collection_factory.go:
```diff
@@ -0,0 +1,17 @@
+
+package collectionfactory
+
+// CollectionFactory is an interface for *descs.CollectionFactory.
+// It's created to avoid cyclic dependency.
+type CollectionFactory interface {
+	ImplmentCollectionFactory()
+}
```
this smells to me. What is the circularity? Can we define narrower interfaces at the relevant call sites?

ZhouXing19 replied:
It's the loop between`cloud` and `descs`.

```
.-> //pkg/cloud:cloud (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog/descs:descs (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog/hydrateddesc:hydrateddesc (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog/lease:lease (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/multitenant:multitenant (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
`-- //pkg/cloud:cloud (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
```

I encountered this because I wanted to have a `CollectionFactory` field in `ExternalStorageContext`, so that we can pass it to `InternalFileToTableExecutor` and use it in `file_table_read_write.go`.

ZhouXing19 replied:
> Can we define narrower interfaces at the relevant call sites?

I'm not sure. In this PR we only refactored the use cases with DDLs. In the future we will want to migrate [all usages](https://docs.google.com/spreadsheets/d/1MQV5IE-01_Owfrp9HIIfXIIERXyTjvzA8oa0SUNBysU/edit?usp=sharing) with not-nil txn to `CollectionFactory.TxnWithExecutor()`. Not sure if we will encounter a larger number and more complex dependency loops.

ZhouXing19 replied:
> I wanted to have a CollectionFactory field in ExternalStorageContext, so that we can pass it to InternalFileToTableExecutor

For more context, I was following how `InternalFileToTableExecutor.ie` gets written to construct the same path for `InternalFileToTableExecutor.cf` .

ajwerner replied:
When a dependency cycle gets in the way, the answer is almost never to smuggle an object through an interface. For a hack, for a brief moment, it can be okay. In some cases, where the idea of opaque dependency injection is deeply fundamental to the problem, then it can be the right thing. In general, it's a code smell. [This PR](https://github.com/cockroachdb/cockroach/pull/87140) should fix the dependency cycle.

ZhouXing19 replied:
I see, thanks for explaining! I just rebased from that PR and it nicely solved the loop. Will think twice when I encounter a dependency loop next time!

ajwerner commented on pkg/sql/internal.go:
```diff
@@ -817,6 +817,11 @@ func (ie *InternalExecutor) execInternal(
 	stmt string,
 	qargs ...interface{},
 ) (r *rowsIterator, retErr error) {
+	if ie.extraTxnState != nil {
+		if txn != ie.extraTxnState.txn {
```
`if ie.extraTxnState != nil && ie.extraTxnState.txn != txn {`

ZhouXing19 replied:
Done.

ajwerner commented on pkg/sql/internal.go:
```diff
@@ -879,6 +884,11 @@ func (ie *InternalExecutor) execInternal(
 	timeReceived := timeutil.Now()
 	parseStart := timeReceived
 	parsed, err := parser.ParseOne(stmt)
+	if tree.CanModifySchema(parsed.AST) && txn != nil {
+		if ie.extraTxnState == nil || ie.extraTxnState.descCollection == nil {
```
I think we need a method on `ie` to detect whether it has a bound transaction or not. All of this nil checking and differentiated nil checking is hard to read. Also, combine the conditions logically as opposed to nesting the if statements.

ZhouXing19 replied:
Done.

ZhouXing19 committed:
sql: migrate `deleteTempTables` to use `planner.WithInternalExecutor()`

This is another DDL statement executed via an internal executor mal-inited.
Change it to use the right interface.

Release justification:
Release note: none.

ZhouXing19 committed:
sql: require txn-related metadata if running DDLs with internal executor with txn

When using internal executor to run DDL statements under a not-nil outer txn,
we require txn-related metadata (such as descriptor collections) to be passed
to the internal executor from the outer caller too. This commit is to add
a gate for this use case.

Release justification: bug fix
Release note: none

ZhouXing19 commented:
Sorry for the noise, was fixing tiny nits in the CI. 
The PR is now ready for another look.

ajwerner commented on pkg/sql/create_external_connection.go:
```diff
@@ -112,21 +114,27 @@ func (p *planner) createExternalConnection(
-	// `system.external_connections` table.
-	if err := ex.Create(params.ctx, params.ExecCfg().InternalExecutor, params.p.User(), p.Txn()); err != nil {
-		return errors.Wrap(err, "failed to create external connection")
+	if err := p.WithInternalExecutor(params.ctx, func(ctx context.Context, txn *kv.Txn, ie sqlutil.InternalExecutor) error {
```
nit: could just be `return p.WithInternalExecutor(...`

ZhouXing19 replied:
Done.

ajwerner commented on pkg/sql/discard.go:
```diff
@@ -84,20 +86,23 @@ func (n *discardNode) startExec(params runParams) error {
-	for _, dbDesc := range allDbDescs {
-		schemaName := p.TemporarySchemaName()
-		err = cleanupSchemaObjects(ctx, p.execCfg.Settings, p.Txn(), descCol, codec, ie, dbDesc, schemaName)
+	if err := p.WithInternalExecutor(ctx, func(ctx context.Context, txn *kv.Txn, ie sqlutil.InternalExecutor) error {
```
nit: could just be `return p.WithInternalExecutor(...`

ZhouXing19 replied:
Done.

ajwerner commented on pkg/sql/internal.go:
```diff
@@ -896,6 +901,10 @@ func (ie *InternalExecutor) execInternal(
 	timeReceived := timeutil.Now()
 	parseStart := timeReceived
 	parsed, err := parser.ParseOne(stmt)
+	if tree.CanModifySchema(parsed.AST) && txn != nil && !ie.checkIfTxnBound() {
```
nit: encapsulate the condition and error into the helper? `if err := ie.checkIfStmtIsAllowed(parsed, txn); err != nil { return nil, err }`?

ZhouXing19 replied:
Done.

ajwerner reviewed (approved):
LGTM. All my comments are minor nits

ZhouXing19 committed:
sql: add a space to a line to fix the lint

It was causing the lint in CI to fail.

Release note: None

ZhouXing19 commented on pkg/sql/internal.go:
```diff
@@ -831,6 +831,11 @@ func (ie *InternalExecutor) execInternal(
 	stmt string,
 	qargs ...interface{},
 ) (r *rowsIterator, retErr error) {
+	if !ie.checkIfTxnIsConsistent(txn) {
```
Done.

rafiss reviewed (approved):
very nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86334#-:-NCMXgEOCYGBInh4pS4X:biuimdb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @shermanCRL, and @ZhouXing19)

___
*[`pkg/ccl/backupccl/backup_test.go` line 8183 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/86334#-NCMH5sY-fS6kBTLjG3M:-NCMH5sY-fS6kBTLjG3N:barkgyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7d881fafc170b75580b1a08ba81a1c38b255b5f/pkg/ccl/backupccl/backup_test.go#L8183)):*
> ```Go
> 		blobs.TestBlobServiceClient(dir),
> 		username.RootUserName(),
> 		nil, /* Internal Executor */
> ```

nit: the convention here (and for all other inline argument comments) would be to use the name of the parameter, but this isn't a huge deal


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
TFTR!
The CI test failure seems flaky.
bors r=ajwerner,rafiss
